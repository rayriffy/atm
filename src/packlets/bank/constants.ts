export enum Note {
  Five = '5',
  Ten = '10',
  Twenty = '20',
}

export enum ErrorCode {
  InsufficientFunds = 'InsufficientFunds',
  InsufficientBanknotes = 'InsufficientBanknotes',
  CombinationNotFound = 'CombinationNotFound',
}
