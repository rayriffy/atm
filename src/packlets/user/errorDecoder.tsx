import { Alert, AlertIcon } from "@chakra-ui/react"

import { decodeError } from "./decodeError"
import { useWithdrawInput } from "./useWithdrawInput"

export const ErrorDecoder = () => {
  const { error } = useWithdrawInput()

  if (error === null) return null

  return (
    <Alert status="error" rounded="md" variant="solid">
      <AlertIcon />
      {decodeError(error)}
    </Alert>
  )
}
