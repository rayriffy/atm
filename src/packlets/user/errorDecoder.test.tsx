import { render } from '@testing-library/react'
import { beforeEach, describe, expect, it, vi } from 'vitest'

import { ErrorDecoder } from './errorDecoder'

import { ErrorCode } from '$bank/constants'


const { useWithdrawInputSpy, decodeErrorSpy } = vi.hoisted(() => ({
  useWithdrawInputSpy: vi.fn<unknown[], { error: ErrorCode | null }>(() => ({
    error: null,
  })),
  decodeErrorSpy: vi.fn().mockReturnValue('mock-error'),
}))

vi.mock('./decodeError', () => ({
  decodeError: decodeErrorSpy,
}))

vi.mock('./useWithdrawInput', () => ({
  useWithdrawInput: useWithdrawInputSpy,
}))

describe('decodeError', () => {
  beforeEach(() => {
    vi.clearAllMocks()
  })

  it('should render nothing if no error', () => {
    useWithdrawInputSpy.mockReturnValue({ error: null })

    const { container } = render(<ErrorDecoder />)

    expect(container.firstChild).toBeNull()
    expect(container).toMatchSnapshot()
  })

  it.each(Object.values(ErrorCode))(
    'should render decoded error (code = %s)',
    errorCode => {
      useWithdrawInputSpy.mockReturnValue({ error: errorCode })
      decodeErrorSpy.mockReturnValue(errorCode)

      const { container, getByText } = render(<ErrorDecoder />)

      expect(decodeErrorSpy).toBeCalledTimes(1)
      expect(decodeErrorSpy).toBeCalledWith(errorCode)
      expect(getByText(errorCode)).toBeDefined()

      expect(container).toMatchSnapshot()
    }
  )
})
