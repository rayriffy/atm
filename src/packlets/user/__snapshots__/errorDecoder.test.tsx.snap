// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`decodeError > should render decoded error (code = CombinationNotFound) 1`] = `
<div>
  <div
    class="chakra-alert css-ivpu1i"
    data-status="error"
    role="alert"
  >
    <span
      class="chakra-alert__icon css-1c6vscd"
      data-status="error"
    >
      <svg
        class="chakra-icon css-qk6lof"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
          fill="currentColor"
        />
      </svg>
    </span>
    CombinationNotFound
  </div>
</div>
`;

exports[`decodeError > should render decoded error (code = InsufficientBanknotes) 1`] = `
<div>
  <div
    class="chakra-alert css-ivpu1i"
    data-status="error"
    role="alert"
  >
    <span
      class="chakra-alert__icon css-1c6vscd"
      data-status="error"
    >
      <svg
        class="chakra-icon css-qk6lof"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
          fill="currentColor"
        />
      </svg>
    </span>
    InsufficientBanknotes
  </div>
</div>
`;

exports[`decodeError > should render decoded error (code = InsufficientFunds) 1`] = `
<div>
  <div
    class="chakra-alert css-ivpu1i"
    data-status="error"
    role="alert"
  >
    <span
      class="chakra-alert__icon css-1c6vscd"
      data-status="error"
    >
      <svg
        class="chakra-icon css-qk6lof"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M11.983,0a12.206,12.206,0,0,0-8.51,3.653A11.8,11.8,0,0,0,0,12.207,11.779,11.779,0,0,0,11.8,24h.214A12.111,12.111,0,0,0,24,11.791h0A11.766,11.766,0,0,0,11.983,0ZM10.5,16.542a1.476,1.476,0,0,1,1.449-1.53h.027a1.527,1.527,0,0,1,1.523,1.47,1.475,1.475,0,0,1-1.449,1.53h-.027A1.529,1.529,0,0,1,10.5,16.542ZM11,12.5v-6a1,1,0,0,1,2,0v6a1,1,0,1,1-2,0Z"
          fill="currentColor"
        />
      </svg>
    </span>
    InsufficientFunds
  </div>
</div>
`;

exports[`decodeError > should render nothing if no error 1`] = `<div />`;
